src/main/java/org/llorllale/cactoos/matchers/HasValues.java

```diff
@@ -53,43 +57,76 @@
  * @checkstyle ProtectedMethodInFinalClassCheck (200 lines)
  */
 public final class HasValues<X> extends TypeSafeDiagnosingMatcher<Iterable<X>> {
+
+    /**
+     * The expected values within the collection.
+     */
+    private final Collection<X> exp;
+
     /**
      * The function to check the {@link Iterable}.
      */
     private final UncheckedFunc<X, Boolean> fnc;
 
+    /**
+     * The function to describe the actual values in hamcrest terms.
+     */
+    private final UncheckedBiProc<Iterable<X>, Description> fact;
+
     /**
      * Ctor.
      * @param exp The expected values within unit test.
-     * @todo #32:30m Add the opportunity to print the exp values to the
-     *  description in order to avoid abstract message like
-     *  "The function applied to X,Y,Z is failed".
      */
     @SafeVarargs
     public HasValues(final X... exp) {
-        this(x -> new ListOf<>(exp).contains(x));
+        this(
+            new ListOf<>(exp),
+            x -> new ListOf<>(exp).contains(x),
+            (act, desc) -> desc.appendValue(new ListOf<>(act))
+        );
     }
 
     /**
      * Ctor.
      * @param fnc The function to check the {@link Iterable}.
      */
     public HasValues(final Func<X, Boolean> fnc) {
+        this(Collections.emptyList(), fnc,
+            (act, desc) -> desc.appendText("The function applied to ")
+                .appendValue(new ListOf<>(act))
+                .appendText(" is failed.")
+        );
+    }
+
+    /**
+     * Ctor.
+     * @param exp The expected values within the unit test.
+     * @param fnc The function to check the {@link Iterable}.
+     * @param fact The function to describe actual values in hamcrest terms.
+     */
+    public HasValues(final Collection<X> exp, final Func<X, Boolean> fnc,
+        final BiProc<Iterable<X>, Description> fact) {
         super();
+        this.exp = exp;
         this.fnc = new UncheckedFunc<>(fnc);
+        this.fact = new UncheckedBiProc<>(fact);
     }
 
     @Override
     public void describeTo(final Description dsc) {
-        dsc.appendText("No description");
+        if (this.exp.isEmpty()) {
```

Comment:
@dgroup shouldn't this check instead be comparing `exp.size()` against the `actual.size()` passed to `matchesSafely`?

Reply:
@llorllale , disagree.
Let's clarify the purpose/goal of the `org.hamcrest.SelfDescribing#describeTo`:
```java
    /**
     * Generates a description of the object.  The description may be part of a
     * a description of a larger object of which this is just a component, so it 
     * should be worded appropriately.
     * 
     * @param description
     *     The description to be built or appended to.
     */
```
Too abstract, let's be more clear, the following test
```java
MatcherAssert.assertThat(
    new ListOf<>(1, 2, 3),
    new HasValues<>(5)
);
```
with wail with this message:
```java
java.lang.AssertionError:
    Expected: <[5]>
    but: <[1, 2, 3]>
```
This message above was assembled in line `6` and `8`
```java
1     public static <T> void assertThat(String reason, T actual, Matcher<? super T> matcher) {
2         if (!matcher.matches(actual)) {
3             Description description = new StringDescription();
4             description.appendText(reason)
5                       .appendText("\nExpected: ")
6                       .appendDescriptionOf(matcher)
7                       .appendText("\n     but: ");
8            matcher.describeMismatch(actual, description);
9          
10          throw new AssertionError(description.toString());
11      }
12   }
```
so, the line `6` (or  SelfDescribing#describeTo) is required to describe the expected object, which should be equal(or whatever condition) to actual object (which we receive during the testing).

In our case the `HasValues` matcher has 2 type of possible expected arguments:
1) HasValues#HasValues(X...)
2) HasValues(org.cactoos.Func<X,java.lang.Boolean>)

In the first case, it's easy to describe the expected object (as for each object in iterable we can call `toString()`). The second case can't be described in such a way, so, I would like to keep this message as `At least one element within the iterable match the function.`. It will allow us for failed tests like
```java
MatcherAssert.assertThat(
    new ListOf<>("book", "phone"),
    new HasValues<>(exp -> exp.startsWith("the"))
);
```
receive the message
```java
java.lang.AssertionError: 
Expected: At least one element within the iterable match the function.
     but: The function applied to <[book, phone]> is failed.
```

I've checked different ways in order to make it simpler but without success. The main problem is a bad design of `org.hamcrest.Matcher#matches` (and its children):
```java
public interface Matcher<T> extends SelfDescribing {

    /**
     * Evaluates the matcher for argument <var>item</var>.
     * <p/>
     * This method matches against Object, instead of the generic type T. This is
     * because the caller of the Matcher does not know at runtime what the type is
     * (because of type erasure with Java generics). It is down to the implementations
     * to check the correct type.
     *
     * @param item the object against which the matcher is evaluated.
     * @return <code>true</code> if <var>item</var> matches, otherwise <code>false</code>.
     *
     * @see BaseMatcher
     */
    boolean matches(Object item);
    
    ...
    void describeMismatch(Object item, Description mismatchDescription);

   ...
    @Deprecated
    void _dont_implement_Matcher___instead_extend_BaseMatcher_();
}

```
For tests like 
```java
MatcherAssert.assertThat(
    new ListOf<>(1, 2, 3),
    new HasValues<>(5)
);
```
the `HasValues` gives
```java
java.lang.AssertionError: 
Expected: <[5]>
     but: <[1, 2, 3]>
```
which is more than enough.

Reply:
@llorllale , what do you think?

Reply:
@dgroup I realized I made a mistake - I shouldn't have accepted `HasValues` as it is because the original requirement (#32) was to have a matcher `HasValues` that accepts *matching criteria* ("endswith", "startswith", etc), but it ended up also *accepting values directly*.

This means this class is doing two things and its implementation looks convoluted as a result. These two things are also accepted *either-or*: users can provide either one, or the other, or both. It's too complicated and a sure sign of code smell and lack of focus.

`HasValues`, to me, should just accept the expected values, while a new `HasValuesMatching` can accept matchers/criteria. `HasValues` could be based off of `HasValuesMatching`.

Reply:
@llorllale, I made mistake too :). 
Agree, let's implement the splitting.